### ISSUE TITLE
Rejecting socket.receive() on receiveTimeout doesn't allow the app to exit

### ISSUE THREAD

**Issue by ezze** [zeromq/zeromq.js#364]
**Describe the bug**

I'm writing Node.js application consisting of few processes:

- single main process sitting in the memory and working like a web server;
- system user's commands that can be run through CLI and exit when they are done.

Commands running through CLI should do their work and:

- if the main process is up then request and receive some information from there and exit;
- if the main process is down then exit as soon as possible.

When the main process is running everything works fine for me. Command process sends a request, receives a response and exits. But unfortunatelly running a command without the main process binded to TCP port results in hanging the application even if I set [receiveTimeout](http://zeromq.github.io/zeromq.js/classes/request.html#receivetimeout) option. Initially I was looking for a way to use [connectTimeout](http://zeromq.github.io/zeromq.js/classes/request.html#connecttimeout) option but didn't get to know how it should work.

[Here](https://stackoverflow.com/q/59029401/506695) is my related question on how to use `connectTimeout`/`receiveTimeout`.

**Reproducing**

A simple code snippet to reproduce the issue:

```javascript
const { Request } = require('zeromq');

async function notify() {
  let parsedResponse;
  try {
    const message = { event: 'hello world' };
    const socket = new Request({ receiveTimeout: 500 });
    socket.connect('tcp://127.0.0.1:33332');
    await socket.send(JSON.stringify(message));
    const response = await socket.receive();
    parsedResponse = JSON.parse(response.toString());
  }
  catch (e) {
    console.error(e);
  }
  return parsedResponse;
}

(async() => {
  const response = await notify();
  if (response) {
    console.log(response);
  }
  else {
    console.log('Nothing is received.');
  }
})();
```

**Expected behavior**

Expected behavior is to catch a rejection of `await socket.receive()` and exit. But the application remains running with the following output:

```
{ [Error: Socket temporarily unavailable] errno: 11, code: 'EAGAIN' }
Nothing is received.
```

Seems that some resources are still in use. Is it a bug or does any solution to solve the issue gracefully already exist?

**Tested on**
 - OS: Kubuntu 18.04
 - ZeroMQ.js version: 6.0.0-beta.4 (installed using `yarn add zeromq@6.0.0-beta.4 --zmq-shared` because `yarn add zeromq@6.0.0-beta.4` didn't work for me).

**Comment by rolftimmermans** [comment]
Thanks for your report, I have it reproduced. I will investigate and report back.

**Comment by rolftimmermans** [comment]
Alright, from what I can see this is expected behaviour.

On line 9 you are sending a message, which is successfully queued by ZeroMQ and will be delivered as soon as the socket connects in from the background I/O thread.

Then on line 10 you are waiting for 500ms for an incoming message, but this fails if nothing is received within that time (regardless of whether a remote socket is connected).

The background I/O thread remains active and will still wait for a remote connection and refuses to exit until the message that was queued in line 9 is delivered.

So far for the explanation, now for a recommendation which options to set.

In order to stop the background I/O thread from keeping the process active until all queued messages are delivered, you can set [`linger: 0`](http://zeromq.github.io/zeromq.js/classes/request.html#linger) on all sockets, or equivalently [`zmq.context.blocky = false`](http://zeromq.github.io/zeromq.js/classes/context.html#blocky) (if libzmq >= 4.2.0). 

However that's probably not really a solution since the queued message is still "reportedly" delivered. So you should sent `immediate: 0`, which means that [messages shall be queued only to completed connections](http://zeromq.github.io/zeromq.js/classes/request.html#immediate).

Now the call to `send()` will block infinitely if no remote peer is connected, but combine it with a [`sendTimeout`](http://zeromq.github.io/zeromq.js/classes/request.html#sendtimeout) and you probably have the behaviour that you're looking for. 

So, to summarise, if I replace 

```
const socket = new Request({ receiveTimeout: 500 });
```

with

```
const socket = new Request({ immediate: true, sendTimeout: 500, receiveTimeout: 500 });
```

then I receive the following output:

```
[Error: Socket temporarily unavailable] { errno: 35, code: 'EAGAIN' }
Nothing is received.
```

and afterwards the process exits.

Hope this helps!

Regarding the issue of installing `yarn add zeromq@6.0.0-beta.4` would you mind opening a separate issue for it?

**Comment by ezze** [comment]
Thanks a lot for your clarifications. I tried both cases: `{ linger: 0 }` and `{ immediate: true, sendTimeout: 500, receiveTimeout: 500 }` on the local machine and everything works. But on another (server) machine only the second one combination of options works. I still can't figure out in details why `linger` option is not good enough for me but anyway I'm happy with proposed solution.

Regarding the issue of installation... Unfortunatelly, I can't reproduce it on Ubuntu machine anymore. But trying to install ZeroMQ on CentOS server I have [another issue with building](https://github.com/zeromq/zeromq.js/issues/365).
